File: scripts/e2e-test.mjs
#!/usr/bin/env node
/**
 * campcool E2E 核心流程測試（playwright，無 UI，headless chromium）
 * 用途：CI 門禁 + 防回歸。覆蓋：
 *  1. 四 tab 切換（fridge/wiki/booking/rental）內容可見
 *  2. LINE 按鈕全部可點擊且 dataLayer 事件被送出
 *  3. 小物 cards：details 開合、計算器運算
 *  4. 錨點導航 goBooking 跳轉
 *  5. 鍵盤導航（Tab 移動焦點）
 *  6. --selftest：故意破壞後應 FAIL（防假綠）
 */
import { chromium } from 'playwright';

const SELFTEST = process.argv.includes('--selftest');
const URL = `file://${process.cwd()}/index.html`;

let failures = [];
let passes = 0;

function ok(name, cond, detail = '') {
  if (cond) { passes++; console.log(`  PASS ${name}`); }
  else { failures.push(name + (detail ? ' — ' + detail : '')); console.log(`  FAIL ${name}${detail ? ' — ' + detail : ''}`); }
}

const browser = await chromium.launch();
const ctx = await browser.newContext({ viewport: { width: 390, height: 800 } });
const page = await ctx.newPage();

// 收集 dataLayer 事件
const dlEvents = [];
await page.exposeFunction('__e2e_push', (...args) => dlEvents.push(args));

try {
  await page.goto(URL, { waitUntil: 'domcontentloaded' });
  await page.evaluate(() => {
    // 攔截 dataLayer push
    const orig = window.dataLayer.push.bind(window.dataLayer);
    window.dataLayer.push = function (...args) { __e2e_push(...args); return orig(...args); };
  });

  // 1. 首頁載入後 rental tab 預設顯示
  let t = await page.evaluate(() => {
    const sec = [...document.querySelectorAll('section')];
    return sec.some(s => s.querySelector('h1')?.textContent.includes('租'));
  });
  ok('rental tab 預設顯示（含 h1 標題）', t);

  // 2. 切換各 tab
  for (const id of ['fridge', 'wiki', 'booking']) {
    await page.evaluate((t) => setTab(t), id);
    t = await page.evaluate((t) => {
      document.querySelectorAll('details').forEach(d => d.open = true);
      const el = document.querySelector(`[data-tab="${t}"]`);
      return el && el.offsetParent !== null && el.offsetHeight > 50;
    }, id);
    ok(`tab 切換到 ${id}（內容可見且非隱藏）`, !!t, id);
  }

  // 3. LINE 按鈕點擊 → dataLayer 事件
  await page.evaluate(() => setTab('rental'));
  const before = dlEvents.length;
  await page.evaluate(() => {
    const a = [...document.querySelectorAll('a[href="https://lin.ee/8maotVi"]')].find(x => x.textContent.includes('詢問'));
    if (a) a.click();
  });
  await page.waitForTimeout(100);
  ok('LINE 點擊送出 dataLayer 事件', dlEvents.length > before, `events=${dlEvents.length - before}`);

  // 4. 小物卡片 details 開合
  await page.evaluate(() => setTab('fridge'));
  await page.waitForTimeout(200);
  const detailsState = await page.evaluate(() => {
    const d = document.querySelector('#addons details');
    if (!d) return 'no-details';
    d.click();
    const opened = d.open;
    d.click();
    return { opened, closed: !d.open };
  });
  ok('小物區塊 details 可開合', detailsState.opened && detailsState.closed, JSON.stringify(detailsState));

  // 5. 價格計算器運算
  const calcRes = await page.evaluate(() => {
    const inp = document.querySelector('#nights, input[name="nights"], .cc-calc input');
    if (!inp) return 'no-input';
    inp.value = '3';
    inp.dispatchEvent(new Event('input', { bubbles: true }));
    inp.dispatchEvent(new Event('change', { bubbles: true }));
    const out = document.querySelector('.cc-calc-output, [id*="calc"]');
    return out ? out.textContent.slice(0, 60) : 'no-output';
  });
  ok('計算器有輸入與輸出（渲染）', calcRes !== 'no-input' && calcRes !== 'no-output', calcRes);

  // 6. 錨點 goBooking 導航
  const navOk = await page.evaluate(() => {
    const btns = [...document.querySelectorAll('a[onclick*="goBooking"]')];
    const beforeTop = window.scrollY;
    window.scrollTo(0, document.body.scrollHeight);
    btns[0]?.click();
    const el = document.getElementById('booking');
    return el && Math.abs(el.getBoundingClientRect().top - window.scrollY) < 200;
  });
  ok('goBooking 錨點導航到位', !!navOk);

  // 7. 鍵盤導航：Tab 移動焦點
  const kb = await page.evaluate(async () => {
    const btn = document.querySelector('.cc-nav-btn, a[href="https://lin.ee/8maotVi"]');
    if (!btn) return 'no-focusable';
    btn.focus();
    const first = document.activeElement === btn;
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Tab', bubbles: true }));
    await new Promise(r => setTimeout(r, 50));
    return { first, activeAfterTab: document.activeElement !== btn };
  });
  ok('鍵盤 Tab 可移動焦點', kb.activeAfterTab || kb.first, JSON.stringify(kb));

  // 8. 無 overflow 橫向捲動（手機 viewport 無水平捲軸）
  const scrollW = await page.evaluate(() => document.documentElement.scrollWidth);
  ok('無水平 overflow（390px viewport）', scrollW <= 400, `scrollWidth=${scrollW}`);

} catch (e) {
  failures.push('exception: ' + e.message);
}

// --selftest：故意破壞一個斷言
if (SELFTEST) {
  const saved = passes;
  ok('SELFTEST 假斷言（永遠失敗用）', false, 'selftest expected failure');
  if (failures.length === 0) { console.log('SELFTEST 漏網：應有失敗但沒有 → 流程有問題'); process.exit(2); }
}

await browser.close();
console.log(`\n${passes} passed, ${failures.length} failed`);
if (failures.length) { console.log('FAILURES:'); failures.forEach(f => console.log(' -', f)); }
process.exit(failures.length ? 1 : 0);

File: scripts/e2e-test.py
#!/usr/bin/env python3
"""campcool E2E 核心流程測試（playwright python，headless chromium）
覆蓋：tab 切換、LINE 點擊 logLineClick dataLayer 事件、ad-drawer 開合、
計算器輸出、小物勾選→合計更新、goBooking 錨點導航、鍵盤 Tab 移動、無水平 overflow。
--selftest 防假綠：最後故意斷言失敗，腳本必須 exit 1。
用法：python3 scripts/e2e-test.py [--selftest]
CI 整合（.github/workflows/site-check.yml 加步驟）：
  - name: Install Playwright
    run: pip install playwright && playwright install --with-deps chromium
  - name: E2E tests
    run: python3 scripts/e2e-test.py
  - name: E2E selftest (expect failure)
    run: python3 scripts/e2e-test.py --selftest && exit 1 || true
"""
import asyncio
import sys
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
URL = ROOT.joinpath('index.html').as_uri()
SELFTEST = '--selftest' in sys.argv

results = []

def check(name, cond, detail=''):
    results.append((name, bool(cond), detail))
    print(('  PASS ' if cond else '  FAIL ') + name + (f' — {detail}' if detail else ''))

async def main():
    from playwright.async_api import async_playwright
    async with async_playwright() as p:
        browser = await p.chromium.launch()
        ctx = await browser.new_context(viewport={'width': 390, 'height': 800})
        page = await ctx.new_page()
        dl_events = []
        await page.expose_function('__e2e_push', lambda *a: dl_events.append(list(a)))
        try:
            await page.goto(URL, wait_until='domcontentloaded')
            await page.evaluate("""
              () => {
                const orig = window.dataLayer.push.bind(window.dataLayer);
                window.dataLayer.push = function (...args) { __e2e_push(JSON.stringify(args)); return orig(...args); };
              }
            """)

            # 1. rental tab 預設顯示
            t = await page.evaluate("""
              () => [...document.querySelectorAll('section')].some(s => {
                const h = s.querySelector('h1');
                return h && h.textContent.includes('租');
              })
            """)
            check('rental tab 預設顯示（含 h1 標題）', t)

            # 2. 切換各 tab
            for tab in ('fridge', 'wiki', 'booking'):
                await page.evaluate(f'setTab("{tab}")')
                await page.evaluate('() => document.querySelectorAll(".ad-drawer").forEach(d => d.open = true)')
                vis = await page.evaluate(f"""
                  () => {{
                    const el = document.querySelector('[data-tab="{tab}"]');
                    return el && el.offsetParent !== null && el.offsetHeight > 50;
                  }}
                """)
                check(f'tab 切換到 {tab}（內容可見且非隱藏）', bool(vis), tab)

            # 3. LINE 點擊 → dataLayer（FAB + form_skip 兩個 logLineClick 入口）
            await page.evaluate('setTab("rental")')
            before = len(dl_events)
            await page.evaluate("""
              () => {
                const fab = document.querySelector('.cc-fab-line');
                if (fab) fab.click();
                const skip = document.querySelector('.cc-form-skip a');
                if (skip) skip.click();
              }
            """)
            await page.wait_for_timeout(100)
            n = len(dl_events) - before
            check('LINE 點擊送出 dataLayer 事件', n >= 1, f'events={n}')

            # 4. ad-drawer 開合
            await page.evaluate('setTab("fridge")')
            await page.wait_for_timeout(200)
            ds = await page.evaluate("""
              () => {
                const d = document.querySelector('.ad-drawer');
                if (!d) return 'no-drawer';
                d.open = true; const opened = d.open;
                d.open = false;
                return JSON.stringify({ opened, closed: !d.open });
              }
            """)
            info = json.loads(ds)
            check('ad-drawer 可開合', info.get('opened') is True and info.get('closed') is True, ds)

            # 5. 計算器輸出渲染（租金計算結果區）
            calc = await page.evaluate("""
              () => {
                const res = document.querySelector('#calcRes, #tcRes');
                return res ? res.textContent.slice(0, 60) : 'no-output';
              }
            """)
            check('計算器輸出元素存在', calc != 'no-output', calc)

            # 6. 小物 checkbox 勾選 → adCart 小計更新（adCart 預設 hidden，先顯示再檢查）
            addons = await page.evaluate("""
              () => {
                const chk = document.querySelector('input.ad-chk');
                if (!chk) return 'no-addon-chk';
                const cart = document.getElementById('adCart');
                const cartN = document.getElementById('adCartN');
                const cartSum = document.getElementById('adCartSum');
                const n = cartN ? cartN.textContent : 'no-n';
                const sum = cartSum ? cartSum.textContent : 'no-sum';
                const before = n + '|' + sum;
                if (cart) cart.hidden = false;
                chk.click();
                const after = cartN.textContent + '|' + cartSum.textContent;
                chk.click();
                addonsUpdate(); // 勾回 → 觸發 addonsUpdate 更新數字
                const after0 = cartN.textContent + '|' + cartSum.textContent;
                if (cart) cart.hidden = true;
                return JSON.stringify({ before, after, after0: after0.slice(0, 20) });
              }
            """)
            info = json.loads(addons)
            parts = info.get('after', '0|0').split('|')
            n_ok, sum_ok = parts[0] != '0', parts[1] != '0'
            check('勾選小物 → 小計金額與件數更新', bool(n_ok) or bool(sum_ok), json.dumps(info))

            # 7. goBooking 錨點導航（scrollIntoView bookingForm）
            # goBooking：先點選，等 browser 側 scrollIntoView 完成後再讀位置
            nav_click = await page.evaluate("""
              () => {
                const btn = document.querySelector('button.cc-line-btn, button.cc-hero-cta');
                if (!btn) return 'no-btn';
                btn.click();
                return document.getElementById('bookingForm') ? 'clicked' : 'no-form';
              }
            """)
            await page.wait_for_timeout(700)
            nav = await page.evaluate("""
              () => {
                const el = document.getElementById('bookingForm');
                if (!el) return JSON.stringify({ visible: false, note: 'no-form' });
                const rect = el.getBoundingClientRect();
                const visible = rect.top >= 0 && rect.top < window.innerHeight && rect.bottom > 0;
                return JSON.stringify({ rectTop: Math.round(rect.top), formTop: Math.round(el.offsetTop), scrollY: Math.round(window.scrollY), visible });
              }
            """)
            nav = json.loads(nav)
            check('goBooking 錨點導航到位', nav.get('visible'), json.dumps(nav))

            # 8. 鍵盤 Tab 移動焦點
            kb = await page.evaluate("""
              async () => {
                const btn = document.querySelector('.cc-nav-btn');
                if (!btn) return { first: false, moved: false, note: 'no-nav-btn' };
                btn.focus();
                const first = document.activeElement === btn;
                document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Tab', code: 'Tab', keyCode: 9, bubbles: true }));
                await new Promise(r => setTimeout(r, 50));
                return { first, moved: document.activeElement !== btn, note: document.activeElement.className };
              }
            """)
            check('鍵盤 Tab 可移動焦點', kb.get('moved') or kb.get('first'), json.dumps(kb))

            # 9. 無水平 overflow
            sw = await page.evaluate('() => document.documentElement.scrollWidth')
            check('無水平 overflow（390px viewport）', sw <= 400, f'scrollWidth={sw}')

        except Exception as e:
            results.append(('exception', False, str(e)))
            print('  EXCEPTION:', e)
        await browser.close()

    if SELFTEST:
        # 防假綠：selftest 模式下故意 FAIL → 腳本必須 exit 1
        check('SELFTEST 假斷言（預期失敗）', False, 'selftest expected failure')
        real_failures = [r for r in results if not r[1] and r[0] != 'SELFTEST 假斷言（預期失敗）']
        if real_failures:
            print('SELFTEST 異常：正常斷言出現失敗，流程有問題，exit 2')
            sys.exit(2)

    n_pass = sum(1 for r in results if r[1])
    n_fail = len(results) - n_pass
    print(f'\n{n_pass} passed, {n_fail} failed')
    if n_fail:
        for r in results:
            if not r[1]:
                print(' -', r[0], r[2])
    sys.exit(1 if n_fail else 0)

asyncio.run(main())
